File: 04/sol.py
def read_file(filename="input.txt"):
    with open("input.txt", "r") as f:
        return f.read().strip()


class Interval:
    def __init__(self, a, b):
        self.start = min(a, b)
        self.end = max(a, b)

    def contains(self, other):
        return self.start <= other.start and other.end <= self.end

    @classmethod
    def any_contain(cls, i1, i2):
        return i1.contains(i2) or i2.contains(i1)

    @classmethod
    def overlap(cls, i1, i2):
        return not (i2.end < i1.start or i1.end < i2.start)


def parse(input_string):
    lines = input_string.split("\n")

    def f(line):
        ivls = line.split(",")
        nums = [tuple(int(i) for i in ivl.split("-")) for ivl in ivls]
        return tuple(Interval(*ivl) for ivl in nums)

    return [f(line) for line in lines]


def solve1(ivls):
    return sum(Interval.any_contain(*ivl) for ivl in ivls)


def solve2(ivls):
    return sum(Interval.overlap(*ivl) for ivl in ivls)


if __name__ == "__main__":
    input_string = read_file("input.txt")
    assert input_string is not None
    ivls = parse(input_string)
    res1 = solve1(ivls)
    print(res1)
    res2 = solve2(ivls)
    print(res2)

File: 04/sol.jl
function readstring(filename)
    input = rstrip(read(filename, String))
    @assert !isempty(input) "File Empty"
    return input
end

function parseinput(input)
    lines = split(input, "\n")
    re = r"^(\d+)-(\d+),(\d+)-(\d+)$"
    function line2ints(line)
        m = match(re, line)
        m = map(n -> parse(Int, n), match(re, line).captures)
        return (m[1], m[2]), (m[3], m[4])
    end
    ints = map(line2ints, lines)
    return ints
end

function contains(i::Tuple{Int, Int}, j::Tuple{Int, Int})::Bool
    return i[1] <= j[1] && j[2] <= i[2]
end

function anycontains(i::Tuple{Int, Int}, j::Tuple{Int, Int})::Bool
    return contains(i, j) || contains(j, i)
end

function overlaps(i::Tuple{Int, Int}, j::Tuple{Int, Int})::Bool
    return !(j[2] < i[1] || i[2] < j[1])
end

function solve1(x)
    return sum(anycontains(i, j) for (i, j) in x)
end

function solve2(x)
    return sum(overlaps(i, j) for (i, j) in x)
end

if abspath(PROGRAM_FILE) == @__FILE__
    filename = isempty(ARGS) ? "input.txt" : ARGS[1]
    input = readstring(filename)
    x = parseinput(input)
    r1 = solve1(x)
    println(r1)
    r2 = solve2(x)
    println(r2)
end
